#! /bin/sh /usr/share/dpatch/dpatch-run
# Description: Fix some compile warnings that prevent building on Wheezy

@DPATCH@
diff -urNad mysql-cluster-7.2.5.orig/scripts/mysqld_safe.sh mysql-cluster-7.2.5/scripts/mysqld_safe.sh
--- mysql-cluster-7.2.5.orig/scripts/mysqld_safe.sh	2012-02-27 12:49:47.651525268 +0000
+++ mysql-cluster-7.2.5/scripts/mysqld_safe.sh	2012-02-27 12:50:31.531526764 +0000
@@ -29,9 +29,8 @@
 syslog_tag_mysqld=mysqld
 syslog_tag_mysqld_safe=mysqld_safe
 
-trap '' 1 2 3 15			# we shouldn't let anyone kill us
 # >> MCP_BUG33984
-trap '' 13                              # not even SIGPIPE
+trap '' 13                              # we shouldn't let SIGPIPE kill us, other signals are handled properly later
 # << MCP_BUG33984
 
 umask 007
@@ -135,7 +134,7 @@
       # sed buffers output (only GNU sed supports a -u (unbuffered) option)
       # which means that messages may not get sent to syslog until the
       # mysqld process quits.
-      cmd="$cmd 2>&1 | logger -t '$syslog_tag_mysqld' -p daemon.error"
+      cmd="$cmd 2>&1 | logger -t '$syslog_tag_mysqld' -p daemon.error & wait"
       ;;
     *)
       echo "Internal program error (non-fatal):" \
@@ -713,6 +712,13 @@
   fi
 fi
 
+# 
+# From now on, we catch signals to do a proper shutdown of mysqld
+# when signalled to do so.
+#
+trap '/usr/bin/mysqladmin --defaults-extra-file=/etc/mysql/debian.cnf refresh & wait' 1 # HUP
+trap '/usr/bin/mysqladmin --defaults-extra-file=/etc/mysql/debian.cnf shutdown' 2 3 15 # INT QUIT and TERM
+
 #
 # Uncomment the following lines if you want all tables to be automatically
 # checked and repaired during startup. You should add sensible key_buffer
